main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -23,6 +23,7 @@ import ca.uwaterloo.flix.language.phase.jvm.JvmOps
 import ca.uwaterloo.flix.language.phase.unification.Unification
 import ca.uwaterloo.flix.util.collection.{ListMap, ListOps}
 import ca.uwaterloo.flix.util.{InternalCompilerException, Result}
+import org.apache.commons.lang3.reflect.MethodUtils
```

Comment:
This is the secret sauce that solves most of our problems. Finally.

https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/MethodUtils.html#getMatchingAccessibleMethod(java.lang.Class,java.lang.String,java.lang.Class...)

Reply:
There is only one concern here:

This method can match primitive parameter by passing in wrapper classes. For example, a [Boolean](https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html) will match a primitive boolean parameter.

Reply:
But that would probably lead us to generate incorrect bytecode.

Reply:
How is there not just a function that does the correct thing, no weird special cases 😄

Reply:
This function from apache is close